### PR TITLE
Make drop logic work better for non-8-step-plants

### PR DESCRIPTION
**Pull request by Thomas--S**
Tested, seems to work fine.
The behavior of 8-step-plants seems to be unchanged.

### CODE REVIEW

File: mods/farming/api.lua [hunk 1]
```diff
@@ -315,10 +315,10 @@ farming.register_plant = function(name, def)
 	for i = 1, def.steps do
 		local drop = {
 			items = {
-				{items = {mname .. ":" .. pname}, rarity = 9 - i},
-				{items = {mname .. ":" .. pname}, rarity= 18 - i * 2},
-				{items = {mname .. ":seed_" .. pname}, rarity = 9 - i},
-				{items = {mname .. ":seed_" .. pname}, rarity = 18 - i * 2},
+				{items = {mname .. ":" .. pname}, rarity = 9 - ( i / 8 ) * def.steps},
```
Comment (C1ffisme) [inline mods/farming/api.lua]: Maybe I'm not quite understanding, but shouldn't this be `rarity = 9 - ( i / def.steps)`

Comment (paramat) [inline mods/farming/api.lua]: Not sure about either yet, still thinking.
When i = 1 rarity should be 8
When i = def.steps rarity should be 1

Comment (Thomas--S) [inline mods/farming/api.lua]: I think this is the curve we want:
https://ggbm.at/rC9aVmZq (Link to online GeoGebra)

Comment (paramat) [inline mods/farming/api.lua]: rarity is linear with i, let's keep it that way.

**Comment by Thomas--S** [comment]
New formula. Tested now.
Look at it in GeoGebra: https://www.geogebra.org/m/xvjRdKC7

**Comment by sofar** [comment]
That looks correct, but it's weird to have `= -` like that :)

**Comment by paramat** [comment]
Not sure about this.
When i = 1 rarity should be 8
When i = def.steps rarity should be 1
With this PR when i = 1 rarity still depends on def.steps, should it be 8 independent of def.steps, so that no matter how many steps a crop has, harvesting on first step has a chance of 1 in 8?

**Comment by sofar** [comment]
test code:

```
local function go(steps)
	for i = 1,steps do
		print(9.0 - i, 18.0 - i * 2, " => ", - ( 8 / steps ) * i + 9, - ( 16 / steps ) * i + 18)
	end
end

go(8)
```

prints out:

```
8.0	16.0	 => 	8.0	16.0
7.0	14.0	 => 	7.0	14.0
6.0	12.0	 => 	6.0	12.0
5.0	10.0	 => 	5.0	10.0
4.0	8.0	 => 	4.0	8.0
3.0	6.0	 => 	3.0	6.0
2.0	4.0	 => 	2.0	4.0
1.0	2.0	 => 	1.0	2.0
```

So it appears to be correct.

for e.g. `5` it prints out:

```
8.0	16.0	 => 	7.4	14.8
7.0	14.0	 => 	5.8	11.6
6.0	12.0	 => 	4.2	8.4
5.0	10.0	 => 	2.6	5.2
4.0	8.0	 => 	1.0	2.0
```

which seems like an improvement - a base chance of 8 seems harsh.

**Comment by paramat** [comment]
Seems ok for trying to harvest a seedling.
With a crop of 2 steps rarities are 5 and 10 which is too generous.
I think better maths can be found.

**Comment by sofar** [comment]
no, a 2 step crop would be:

```
8.0	16.0	 => 	5.0	10.0
7.0	14.0	 => 	1.0	2.0
```

which you should read as follows:

first step plant was 8,16 is now 5,10 
second step of plant was 7,14 is now 1,2

so with this change the full grown 2 step plant has a 100% drop chance and another 50% for one more item.

Still seems reasonable to me

**Comment by sofar** [comment]
oh, misread, you said "too generous".

Welp, idk. 20% isn't very high

**Comment by paramat** [comment]
8 - (i - 1) / (def.steps - 1) * 7
Then double this value for the following line.

**Comment by jhcole** [comment]
I agree with paramat's rarity formula, but whatever formula is chosen, I suggest it be written in this form:
```
local base_rarity = 8 - (i-1)*7/(def.steps - i)
    local drop = {
      items = {
        {items = {mname .. ":" .. pname}, rarity = base_rarity},
        {items = {mname .. ":" .. pname}, rarity = base_rarity * 2},
        {items = {mname .. ":seed_" .. pname}, rarity = base_rarity},
        {items = {mname .. ":seed_" .. pname}, rarity = base_rarity * 2},
      }
    }
```

**Comment by sofar** [comment]
ohh, that looks way better, yes, please.

**Comment by paramat** [comment]
Yes perfect.

**Comment by Thomas--S** [comment]
The problem of paramat's formula is, that it doesn't work for 1-step-plants.

**Comment by paramat** [comment]
1 step crops won't be used.

**Comment by Thomas--S** [comment]
Example:

My formula:

Step|8-step plant||Step|4-step plant
--|--|--|--|--
1|0.13
2|0.14 |≙|1|0.14
3|0.17
4|0.2 |≙|2|0.2
5|0.25
6|0.33 |≙|3|0.33
7|0.5
8|1 |≙|4|1

Paramat's formula:

Step|8-step plant||Step|4-step plant
--|--|--|--|--
1|0.13
2|0.14 |≙|1|0.13
3|0.17
4|0.2 |≙|2|0.18
5|0.25
6|0.33 |≙|3|0.3
7|0.5
8|1 |≙|4|1

So, actually I still prefer my formula, because step one of a 4-step-plant actually corresponds to step 2 of an 8-step-plant. Seen from this point of view, the values of my formula match at every step, whereas the values of paramat's formula only match at step 1 and the last step.

However, it is still fine for me to add paramat's formula, it only seemed a little bit strange to me.

**Comment by paramat** [comment]
> step one of a 4-step-plant actually corresponds to step 2 of an 8-step-plant. Seen from this point of view, the values of my formula match at every step,

That's irrelevant. What matters is that rarity at first step is always 8, at last step is 1, and varies linearly between.

> With a crop of 2 steps rarities on first step are 5 and 10 which is too generous.

**Comment by jhcole** [comment]
Maybe I'm over thinking this, but the two proposed formulas can be related by adding a scalar `a` to each term.
```
base_rarity = 8 + a - (i-1+a)*(7+a)/(def.steps-1+a)
```
If we choose `a=1` then we have Thomas--S' formula. If we choose `a=0` we have paramat's formula. We can avoid the divide by zero problem in paramat's with any choice of `a` such that `a!=0`. For example `a=0.01`.

**Comment by paramat** [comment]
Well, we don't need an option of formulas, however that divide by zero issue should be avoided in the code, thanks for spotting that.

**Comment by paramat** [comment]
Or better, allow 1 step plants but do 'if def.steps == 1 then baserarity = 1 else my formula'

**Comment by Thomas--S** [comment]
I have updated the PR now:
It still uses the old formula, but transposed, so that there is one operation less.
It sill uses the old formula, because I think the benefits of @paramat's suggestions aren't worth its complexity.

**Comment by paramat** [comment]
:-1: 

Two of us like this https://github.com/minetest/minetest_game/pull/1406#issuecomment-263101858
as long as it catches the divide by zero problem this way https://github.com/minetest/minetest_game/pull/1406#issuecomment-263765032

**Comment by paramat** [comment]
Needs adoption and a new PR.

**Comment by Thomas--S** [comment]
I already have an idea for a totally new formula. I still have to find the time to find the mathematical expression :)

**Comment by paramat** [comment]
Ok, whatever happens, feel free to submit it, even if our preferred code above goes ahead temporarily.

**Comment by C1ffisme** [comment]
Little nitpick, but grammatically shouldn't it be 'Make drop logic **work** better for non-8-step-plants'?

**Comment by Thomas--S** [comment]
Replaced by #1544

**Pull request closed without merging**